num_enum_derive/src/lib.rs

```diff
@@ -145,9 +358,116 @@ pub fn derive_into_primitive(input: TokenStream) -> TokenStream {
     })
 }
 
+/// Implements `From<Primitive>` for a `#[repr(Primitive)] enum`.
+///
+/// Turning a primitive into an enum with `from`.
+/// ----------------------------------------------
+///
+/// ```rust
+/// use num_enum::FromPrimitive;
+///
+/// #[derive(Debug, Eq, PartialEq, FromPrimitive)]
+/// #[repr(u8)]
+/// enum Number {
+///     Zero,
+///     #[num_enum(default)]
+///     NonZero,
+/// }
+///
+/// fn main() {
+///     let zero = Number::from(0u8);
+///     assert_eq!(zero, Number::Zero);
+///
+///     let one = Number::from(1u8);
+///     assert_eq!(one, Number::NonZero);
+///
+///     let two = Number::from(2u8);
+///     assert_eq!(two, Number::NonZero);
+/// }
+/// ```
+#[proc_macro_derive(FromPrimitive, attributes(num_enum))]
+pub fn derive_from_primitive(input: TokenStream) -> TokenStream {
+    let enum_info: EnumInfo = parse_macro_input!(input);
+    let krate = Ident::new(&get_crate_name(), Span::call_site());
+
+    let default_ident: Ident = match enum_info.default() {
+        Some(ident) => ident.clone(),
+        None => {
+            let span = Span::call_site();
+            let message =
+                "#[derive(FromPrimitive)] requires a variant marked with `#[num_enum(default)]`";
+            return syn::Error::new(span, message).to_compile_error().into();
+        }
+    };
+
+    let EnumInfo {
+        ref name, ref repr, ..
+    } = enum_info;
+
+    let variant_idents: Vec<Ident> = enum_info.variant_idents();
+    let expression_idents: Vec<Vec<Ident>> = enum_info.expression_idents();
+    let variant_expressions: Vec<Vec<Expr>> = enum_info.variant_expressions();
+
+    debug_assert_eq!(variant_idents.len(), variant_expressions.len());
+
+    TokenStream::from(quote! {
+        impl ::#krate::FromPrimitive for #name {
+            type Primitive = #repr;
+
+            fn from_primitive(number: Self::Primitive) -> Self {
+                // Use intermediate const(s) so that enums defined like
+                // `Two = ONE + 1u8` work properly.
+                #![allow(non_upper_case_globals)]
+                #(
+                    #(
+                        const #expression_idents: #repr = #variant_expressions;
+                    )*
+                )*
+                #[deny(unreachable_patterns)]
+                match number {
+                    #(
+                        #( #expression_idents )|*
+                        => Self::#variant_idents,
+                    )*
+                    #[allow(unreachable_patterns)]
```

Comment:
Sorry, one last question - what is this allow here for (and the one about 100 lines later)? I tried changing them to `deny`, and no tests started failing... Is this being defensive against the compiler getting smarter? Or is it guarding against something we run into today?

Reply:
The match arms could already be exhaustive, in which case the `_ =>` is redundant and thus cannot be reached. Since we have to consider match arms to be undecidable we have to assume them to be non-exhaustive.

This compiles:
```rust
enum Foo {
    Bar,
    Baz,
    Blee,
}

fn main() {
    let foo = Foo::Bar;
    
    #[deny(unreachable_patterns)]
    match foo {
        Foo::Bar => {},
        Foo::Baz => {},
        Foo::Blee => {},
    }
}
```

Now add a `_ => {}` and it fails to compile with:

```rust
error: unreachable pattern
  --> Untitled.rs:16:9
   |
16 |         _ => {},
   |         ^
   |
note: the lint level is defined here
  --> Untitled.rs:10:12
   |
10 |     #[deny(unreachable_patterns)]
   |    
```

Add an `#[allow(unreachable_patterns)]` for it and it compiles again.

An alternative to this would be to check if a variant is marked as default and omit its explicit match arm, in favor of the corresponding `_ =>`, which is a superset of it anyway.

It needs a bit more logic though. I would be fine with either. What's your view on this, @danielhenrymantilla ?

You would basically have to refactor these methods into using a `filter_map` with an additional `if variant.is_default { return None; }`:

```rust
fn variant_idents(&self) -> Vec<Ident>;
fn expression_idents(&self) -> Vec<Vec<Ident>>;
fn variant_expressions(&self) -> Vec<Vec<Expr>>;
```

… and rename them slightly, probably.

Reply:
Do you know why https://github.com/illicitonion/num_enum/pull/23/files#diff-3b0ef9c7bb08eebdf580ebcbdfb3ce98 passes when this is set to deny? I would expect it to fail?

Reply:
Hm, I took the output of `tests/try_build/pass/exhaustive_enum.rs` with `cargo expand`, did some cleanup to make it compile, removed the `#[allow(unreachable_patterns)]` (or changed it to `deny`) and it fails (as expected):

```rust
error: unreachable pattern
   --> dummy/src/main.rs:425:13
    |
425 |             _ => Self::A,
    |             ^
    |
note: the lint level is defined here
   --> dummy/src/main.rs:358:16
    |
358 |         #[deny(unreachable_patterns)]
    |                ^^^^^^^^^^^^^^^^^^^^
```

Did you run `$ cargo clean` to make sure it's not re-using an older build of the proc-macro crate?

Reply:
Weird...! I repro the error when I manually expand the code, but not when I don't... I guess maybe there's some exception for generated code somewhere? Oh well!